### ISSUE TITLE
Date incorrect on pdf invoice

### ISSUE THREAD

**Issue by devchris79** [magento/magento2#26675]
### Preconditions (*)

1. Magento 2.4-develop
2. 

### Steps to reproduce (*)

1. Go to Sales->Orders, select an order and then try to "print invoices"

### Expected result (*)

1. Date to reflect purchase date

### Actual result (*)

1. Dates are incorrect, we have 2006 and 2008 for orders, before the website was developed....
2. Dates were fine in 2.3.2. 
[#26665issueInvoice.pdf](https://github.com/magento/magento2/files/4152912/26665issueInvoice.pdf)

**Comment by anvanza** [comment]
Same issue here but my dates are way higher.
<img width="212" alt="Screenshot 2020-02-03 at 14 01 43" src="https://user-images.githubusercontent.com/2848347/73655281-f00df100-468d-11ea-911d-cf540f2666db.png">

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Delta 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-31082` were created

**Issue Available**: @engcom-Delta, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by devchris79** [comment]
@engcom-Delta Why did you edit the post to include a precondition of Locale Storeview: nl_NL?
This is not a precondition that we used.....

It looks like the date is inserted in:
/vendor/magento/module-sales/Model/Order/Pdf/AbstractPdf.php

So this may affect other pdf documents like shipping, credit memo etc.

**Comment by devchris79** [comment]
Credit memos are indeed also affected.

**Comment by devchris79** [comment]
I think this may be because of the changes in #23693. @edenduong does that sound likely?

**Comment by edenduong** [comment]
Hi @devchris79 ,

Thank you for your issue report. I try to reproduce it on my local site but I can not reproduce  in my site. In my commit, I try to change the logic of `scopeDate similar to the `date` function. Before that, they are different and the scopeDate didn't support full type of interface.

**Comment by engcom-Delta** [comment]
@devchris79 According your steps as its are now, issue is not reproducible. [comment](https://github.com/magento/magento2/issues/26675#issuecomment-581405573) adds more details to haw reproduce issue and  **Locale=Dutch (Netherlands)** is important thing to reproduce issue

**Comment by anvanza** [comment]
@magento give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @anvanza. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by devchris79** [comment]
Hi @edenduong 

Why have you closed this issue? Its been seen by myself, @anvanza  and @engcom-Delta 

Here is our locale options:

![locale](https://user-images.githubusercontent.com/43850303/73754130-cf13d180-475b-11ea-8dae-e9bee8af584c.jpg)

**Comment by edenduong** [comment]
@devchris79 : sorry for my mistake. I didn’t mean that. I reopened it.

**Comment by devchris79** [comment]
@edenduong No problem :-) Maybe the issue is related to the locale option giving the strange results. It seems @engcom-Delta and @anvanza believe this to be pertinent.

**Comment by engcom-Delta** [comment]
@magento give me 2.4-develop instance

**Comment by magento-engcom-team** [comment]
Hi @engcom-Delta. Thank you for your request. I'm working on Magento 2.4-develop instance for you

**Comment by edenduong** [comment]
> @edenduong No problem :-) Maybe the issue is related to the locale option giving the strange results. It seems @engcom-Delta and @anvanza believe this to be pertinent.

I reproduce the issue in my local. Thank you @devchris79  :)

**Comment by magento-engcom-team** [comment]
Hi @engcom-Delta, here is your [Magento instance](https://i-26675-2-4-develop.instances.magento-community.engineering).
Admin access: [https://i-26675-2-4-develop.instances.magento-community.engineering/admin_583c](https://i-26675-2-4-develop.instances.magento-community.engineering/admin_583c)
Login: `8d54cec2` Password: `2a124ed93fd9`
Instance will be terminated in up to 3 hours.

**Comment by engcom-Delta** [comment]
@devchris79 This [case](https://github.com/magento/magento2/issues/26675#issuecomment-581939462) is not reproducible for me

**Comment by edenduong** [comment]
@devchris79: Change source code to :  
```

    public function scopeDate($scope = null, $date = null, $includeTime = false)
    {
        $timezone = new \DateTimeZone(
            $this->_scopeConfig->getValue($this->getDefaultTimezonePath(), $this->_scopeType, $scope)
        );
        switch (true) {
            case (empty($date)):
                $date = new \DateTime('now', $timezone);
                break;
            case ($date instanceof \DateTime):
            case ($date instanceof \DateTimeImmutable):
                $date = $date->setTimezone($timezone);
                break;
            case (is_numeric($date)):
                $date = new \DateTime('@' . $date);
                $date = $date->setTimezone($timezone);
                break;
            default:
                $date = new \DateTime($date, $timezone);
                break;
        }

        if (!$includeTime) {
            $date->setTime(0, 0, 0);
        }

        return $date;
    }
```

will solve this problem. I will make PR.

**Comment by devchris79** [comment]
@edenduong Thanks for the fix, I have applied this and can confirm it works a treat.

**Comment by gwharton** [comment]
Confirmed fix working on 2.3.4 with en_GB system locale.

**Comment by edenduong** [comment]
I just created Pull Request. Thank you.

**Comment by ajwalks** [comment]
Hi, 2.3.4 with same issue. 
Where to put this fix please. 
Is it addition or change existing code?

**Comment by devchris79** [comment]
Hi @ajwalks,

You need to change timezone.php, its at path:
/vendor/magento/framework/Stdlib/DateTime/Timezone.php

Update with the changes from @edenduong and recompile.

**Comment by gwharton** [comment]
If you havent done so already, look into one of the composer patch management packages, then download the pull request as a diff, and apply via your main composer file. I think applying patches like this is preferred to editing core.

**Comment by devchris79** [comment]
It looks like the pull request has issues and has not currently passed the checks, I assume that method would only work on the pull requests that are complete?

**Comment by ajwalks** [comment]
@devchris79, thank you, ace.
@gwharton, will look into this thanks. A new process for me to get my head around that so like most things magento would take me the rest of the day!

**Comment by gwharton** [comment]
> It looks like the pull request has issues and has not currently passed the checks, I assume that method would only work on the pull requests that are complete?

The errors in the Pull request seem like an issue with the testing framework at the moment. They seem to be occurring randomly on other PR's too. If they are re-run, they will probably pass on the second run. I'll trigger a re-run.

**Comment by ajwalks** [comment]
Fix worked a treat, thank you to all involved.

**Comment by devchris79** [comment]
Closing as @edenduong fix and pull request has sorted the issue.

**Comment by gwharton** [comment]
After several reruns, all tests on the PR now pass.

**Comment by hostep** [comment]
@devchris79: please don't close issues when the PR has not been merged yet (if for some reason the PR gets declined and closed, we wouldn't have an open issue anymore to track this bug).

Thanks!

**Comment by magento-engcom-team** [comment]
Hi @devchris79. Thank you for your report. 
The issue has been fixed in magento/magento2#26701 by @edenduong in 2.4-develop branch
Related commit(s):
 - [5ebb43bb59b97f5315ae6421fe3c1b04c6affc70](https://github.com/magento/magento2/commit/5ebb43bb59b97f5315ae6421fe3c1b04c6affc70)

The fix will be available with the upcoming 2.4.0 release.

**Comment by luzat** [comment]
Won't this fix be applied to 2.3?

**Comment by gwharton** [comment]
I don't know what Magento's plan is for backporting to 2.3. Presumably they are still going to be releasing 2.3.5 etc... so I would assume they will be handling the backporting internally, as we are no longer able to generate backports outside of the magento organisation.

In the meantime....

For each issue I care about, that has been merged into 2.4, I create a diff and apply it using the cweagans patching process to my 2.3 installation. At each version release of 2.3, I check each one in turn and if it has been backported into 2.3 by magento, then I remove my local patch. This way, I should be covered in both scenareos.

**Comment by glastron174** [comment]
> 
> 
> If you havent done so already, look into one of the composer patch management packages, then download the pull request as a diff, and apply via your main composer file. I think applying patches like this is preferred to editing core.
> 
> `composer require cweagans/composer-patches `
> Add section to bottom of main composer.json file
> 
> ```
>     "extra": {
>         "magento-force": "override",
>         "composer-exit-on-patch-failure": true,
>         "patches": {
>             "magento/framework": {
>                 "1.0": "patches/composer/26701.diff"
>             }
>         }
>     }
> ```
> 
> Download the patch from https://patch-diff.githubusercontent.com/raw/magento/magento2/pull/26701.diff and save into patches/composer folder
> 
> Edit the patch file and remove all "lib/internal/Magento/Framework/" from paths. so paths would be "a/Stdlib/DateTime/Timezone.php" and "b/Stdlib/DateTime/Timezone.php" etc.
> 
> Now just do a
> 
> `composer install `
> 
> When a new release is out which already incorporates the fix, then remove the patch from your composer file.

Hi, Should the /patches/composer folder created in the root?
I'm getting an error on the composer install 

[Seld\JsonLint\ParsingException]
  "./composer.json" does not contain valid JSON
  Parse error on line 133:
    }
  --------------------^
  Expected one of: '}', ','

Thanks in advanced, Edwin

**Comment by gwharton** [comment]
Looks like there is some whitespace on the end of line 133 that it doesn't like. There shouldnt be any whitespace at the ends of the lines in a json file. Probably just copy and pasted across.

**Comment by gwharton** [comment]
If it is working right, on composer install it should remove framework, reinstall framework and then apply the patch. Then you're good to go. Once the patch is integrated into mainline, you can remove it from your composer.json when you update.

**Comment by glastron174** [comment]
> 
> 
> If it is working right, on composer install it should remove framework, reinstall framework and then apply the patch. Then you're good to go. Once the patch is integrated into mainline, you can remove it from your composer.json when you update.

Thanks, I had to add one } extra in the sjon file. The dates in PDFs are correct now.

**Comment by tushhan** [comment]
I am using Magento 2.3.4

I have applied @edenduong' code to the file located in; /vendor/magento/framework/Stdlib/DateTime/Timezone.php (line 196)

###############

public function scopeDate($scope = null, $date = null, $includeTime = false)
    {
        $timezone = new \DateTimeZone(
            $this->_scopeConfig->getValue($this->getDefaultTimezonePath(), $this->_scopeType, $scope)
        );
        switch (true) {
            case (empty($date)):
                $date = new \DateTime('now', $timezone);
                break;
            case ($date instanceof \DateTime):
            case ($date instanceof \DateTimeImmutable):
                $date = $date->setTimezone($timezone);
                break;
            default:
                $date = new \DateTime(is_numeric($date) ? '@' . $date : $date);
                $date->setTimezone($timezone);
                break;
        }

        if (!$includeTime) {
            $date->setTime(0, 0, 0);
        }

        return $date;
    }

###################

This corects dates on invoices.

**Comment by dudzio12** [comment]
If you use Magento with composer, there is a diff for you:
https://pastebin.com/YtzX09gL

**Comment by JustICTNL** [comment]
@dudzio12 
How can I apply this patch with composer?

> If you use Magento with composer, there is a diff for you:
> https://pastebin.com/YtzX09gL

**Comment by dudzio12** [comment]
@JustICTNL:
1. `composer require cweagans/composer-patches`
2. Create root subdirectory called `patches` and create file `datetime_format_2_3_5.patch. 
3. Add pastebin content to this file
4. Add to your composer.json:
```
(...)
 "extra": {
        "magento-force": "override",
        "enable-patching": true,
        "composer-exit-on-patch-failure": true,
        "patches": {
            "magento/framework": {
                "Datetime format": "patches/datetime_format_2_3_5.patch"
            }
        }
    }
}
```

**Comment by MarcusWolschon** [comment]
> 
> 
> @JustICTNL:
> 
>     1. `composer require cweagans/composer-patches`
> 
>     2. Create root subdirectory called `patches` and create file `datetime_format_2_3_5.patch.
> 
>     3. Add pastebin content to this file
> 
>     4. Add to your composer.json:
> 
> 
> ```
> (...)
>  "extra": {
>         "magento-force": "override",
>         "enable-patching": true,
>         "composer-exit-on-patch-failure": true,
>         "patches": {
>             "magento/framework": {
>                 "Datetime format": "patches/datetime_format_2_3_5.patch"
>             }
>         }
>     }
> }
> ```

Do we have to remove this again before upgrading to the next Magento release?
What steps are required (I have never worked with patches in Composer).

**Comment by dudzio12** [comment]
You probably should delete it as soon as the new release implementing proper fix will be available.

It's good practise to name the patch with issue number just to not get lost in the future:
```
"magento/framework": {
  "Datetime format - Magento 2.3.5 - issue 26675": "patches/datetime_format_2_3_5_26675.patch"
 }
```

**Comment by hostep** [comment]
@MarcusWolschon: with `composer-exit-on-patch-failure` set to `true`, composer will complain if the patch is no longer applicable on a certain version of Magento, which is very handy.
But we always double check all our patches after every new Magento version anyways to make sure they are still needed or not (sometimes a fix is done in a completely different way than the patch and composer wouldn't complain in that case, but that's not happening very often)

**Comment by JustICTNL** [comment]
@dudzio12 and @MarcusWolschon 
I did what you said, but it doesn't work for me :'(

Did I do something wrong, or is there another solution? I had the order date 9-11-2022. In magento is the correct order date.

**Comment by JustICTNL** [comment]
Magento directory/composer.json
`"extra": {
        "magento-force": "override",
        "enable-patching": true,
        "composer-exit-on-patch-failure": true,
        "patches": {
            "magento/framework": {
                "Datetime format": "patches/datetime_format_2_3_5.patch"
            }
        },
`

DIFF file pasted in file:
Magento Directory/patches/datetime_format_2_3_5.patch

**Comment by JustICTNL** [comment]
> I am using Magento 2.3.4
> 
> I have applied @edenduong' code to the file located in; /vendor/magento/framework/Stdlib/DateTime/Timezone.php (line 196)
> 
> ###############
> 
> public function scopeDate($scope = null, $date = null, $includeTime = false)
> {
> $timezone = new \DateTimeZone(
> $this->_scopeConfig->getValue($this->getDefaultTimezonePath(), $this->_scopeType, $scope)
> );
> switch (true) {
> case (empty($date)):
> $date = new \DateTime('now', $timezone);
> break;
> case ($date instanceof \DateTime):
> case ($date instanceof \DateTimeImmutable):
> $date = $date->setTimezone($timezone);
> break;
> default:
> $date = new \DateTime(is_numeric($date) ? '@' . $date : $date);
> $date->setTimezone($timezone);
> break;
> }
> 
> ```
>     if (!$includeTime) {
>         $date->setTime(0, 0, 0);
>     }
> 
>     return $date;
> }
> ```
> 
> ###################
> 
> This corects dates on invoices.

This works now for me. i updgraded to Magento 2.3.5-P1 and now it works.
FInd the file with command:
find . -name Timezone.php
Result: ./vendor/magento/framework/Stdlib/DateTime/Timezone.php

I changed above in ./vendor/magento/framework/Stdlib/DateTime/Timezone.php
And now it works :)

**Comment by IbrahimS2** [comment]
@magento-engcom-team It still persisting even on M2.3.5-p2, can we have this ported/merged to p3 maybe?

**Comment by VincentMarmiesse** [comment]
This issue still exists in version 2.3.6.

**Comment by FredericMartinez** [comment]
Patch from https://github.com/magento/magento2/commit/5ebb43bb59b97f5315ae6421fe3c1b04c6affc70.patch do the job (on 2.3.5-p2)

**Comment by LaurensGoedel** [comment]
Just updated to 2.3.7 and issue still extists

**Comment by hostep** [comment]
@LaurensGoedel: the issue got fix in Magento 2.4.0 and above, it looks like Magento core devs didn't seem to think it was necessary to backport this fix to the 2.3.x release line. If you still need a fix for 2.3.x,  please read through the entire thread, there are plenty of comments that tell you how to apply patches to resolve this issue.

**Comment by Aquive** [comment]
Still exists in M2.3.7. I used this fix: https://stackoverflow.com/questions/64553446/magento-2-3-4-wrong-order-date-in-invoice-pdf
